File: 프로그래머스/lv2/60057. 문자열 압축/문자열 압축.js
function solution(s) {
  const MAX_UNIT_LEN = Math.floor(s.length / 2);
  let minCompressedLen = s.length;

  for (let unitLen = 1; unitLen <= MAX_UNIT_LEN; unitLen += 1) {
    minCompressedLen = Math.min(minCompressedLen, compressedLen(s, unitLen));
  }

  return minCompressedLen;
}

/**
 *
 * @param {string} toCompress
 * @param {number} unitLen
 * @returns {number}
 */
function compressedLen(toCompress, unitLen) {
  let unit = toCompress.slice(0, unitLen);
  let count = 1;
  let ret = 0;

  for (let i = unitLen; unit.length !== 0; i += unitLen) {
    const nextUnit = toCompress.slice(i, i + unitLen);
    if (unit !== nextUnit) {
      ret += unit.length;
      if (count > 1) {
        ret += count.toString().length;
        count = 1;
      }
      unit = nextUnit;
    } else {
      count += 1;
    }
  }
  return ret;
}

File: 프로그래머스/lv2/60057. 문자열 압축/문자열 압축.py
from functools import reduce


def compressed_len(to_compress: str, unit_len: int) -> int:
    unit = to_compress[:unit_len]
    count = 1
    ret = 0

    i = unit_len
    while len(unit) != 0:
        next_unit = to_compress[i:i + unit_len]
        if unit != next_unit:
            ret += len(unit)
            if count > 1:
                ret += len(str(count))
                count = 1
            unit = next_unit
        else:
            count += 1
        i += unit_len

    return ret


def solution(string: str) -> int:
    return reduce(
        lambda min_len, unit: min(min_len, compressed_len(string, unit)),
        range(1, 1 + (len(string) // 2)),
        len(string)
    )
